# Task 4.7: Deriving New Variables

This script includes the following points:

Step 1.1. Create “price_label” column.

Step 1.2. Create "busiest_day" column.

Step 2. Update "busiest day" column to "busiest days" to identify order day of the week as "Busiest days," "Slowest days," or "Regularly busy."

Step 3. Check values of new "busiest days" column for accuracy with observations in markdown format.

Step 4. Create new column "busiest_period_of_day" to identify time periods “Most orders,” “Average orders,” and “Fewest orders.”

Step 5. Print frequency of "busiest_period_of_day" column.

Step 7. Export dataframe as a pickle file to “Prepared Data” folder.

In [5]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

In [6]:
# Importing dataframe
path = r'C:\Users\Sanket\Documents\MM-2020 Instacart Basket Analysis'

In [8]:
#Importing data set ords_prods_merge.pkl
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data','Prepared data','ords_prods_merge.pkl'))

In [15]:
# create a subset
df = df_ords_prods_merge[:1000000]

In [11]:
df

product_id                product_name  aisle_id  department_id  \
0                1  Chocolate Sandwich Cookies        61             19   
1                1  Chocolate Sandwich Cookies        61             19   
2                1  Chocolate Sandwich Cookies        61             19   
3                1  Chocolate Sandwich Cookies        61             19   
4                1  Chocolate Sandwich Cookies        61             19   
...            ...                         ...       ...            ...   
999995        2078  Unsalted Pure Irish Butter        36             16   
999996        2078  Unsalted Pure Irish Butter        36             16   
999997        2078  Unsalted Pure Irish Butter        36             16   
999998        2078  Unsalted Pure Irish Butter        36             16   
999999        2078  Unsalted Pure Irish Butter        36             16   

        prices  order_id  user_id eval_set  order_number  order_dow  \
0          5.8   3139998      138    prior            28          6   
1          5.8   1977647      138    prior            30          6   
2          5.8    389851      709    prior             2          0   
3          5.8    652770      764    prior             1          3   
4          5.8   1813452      764    prior             3          4   
...        ...       ...      ...      ...           ...        ...   
999995    13.5   2280887   102527    prior            25          6   
999996    13.5   3310791   102530    prior             1          4   
999997    13.5   1704867   102530    prior             2          2   
999998    13.5    629818   102530    prior             3          6   
999999    13.5   2328893   102544    prior             5          5   

        order_hour_of_day  days_since_prior_order  add_to_cart_order  \
0                      11                     3.0                  5   
1                      17                    20.0                  1   
2                      21                     6.0                 20   
3                      13                     NaN                 10   
4                      17                     9.0                 11   
...                   ...                     ...                ...   
999995                 14                     6.0                  1   
999996                 17                     NaN                 13   
999997                 17                    30.0                  4   
999998                 18                     4.0                  3   
999999                  9                     1.0                  1   

        reordered _merge  
0               0   both  
1               1   both  
2               0   both  
3               0   both  
4               1   both  
...           ...    ...  
999995          1   both  
999996          0   both  
999997          1   both  
999998          1   both  
999999          0   both  

[1000000 rows x 15 columns]

In [16]:
df.shape

(1000000, 15)

#define your price_label function and apply it to the subset you just created:

# Task Step 1.1- Create “price_label” column.

In [12]:
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [24]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    652638
Low-range product    338018
High range             9344
Name: count, dtype: int64

In [27]:
df['price_range']= df.apply(price_label,axis=1)

C:\Users\Sanket\AppData\Local\Temp\ipykernel_20068\1319572896.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range']= df.apply(price_label,axis=1)


In [28]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    652638
Low-range product    338018
High range             9344
Name: count, dtype: int64

In [29]:
df['prices'].max()

24.5

In [71]:
# Use loc() method on df subset to apply price range function.

df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [31]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [32]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [33]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     652638
Low-range product     338018
High-range product      9344
Name: count, dtype: int64

In [72]:
# Use loc() method on entire ords_prods_merge dataframe to apply price range function.

df_ords_prods_merge.loc[df_ords_prods_merge['prices']<15,'price_range_loc'] = 'High-range product'

In [39]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 


In [40]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [41]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
High-range product    21119602
Low-range product     10126321
Mid-range product       652638
nan                     506298
Name: count, dtype: int64

# Step 1.2. Create a "busiest_day" column.

In [46]:
# Change "order_dow" column name to "order_day_of_week."

df_ords_prods_merge.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [47]:
# Find frequency of orders_day_of_week.

df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [48]:
# Use for-loop to determine whether orders are on "busiest day" (0 = Saturday), "least busy" (4 = Wednesday), or "regularly busy" (other days of the week).

result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [49]:
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Bus

In [50]:
# Create new column in ords_prods_merge to view results in context.

df_ords_prods_merge['busiest_day'] = result

In [51]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

# Step 2. Update "busiest day" column to "busiest days" to identify order day of the week as "Busiest days," "Slowest days," or "Regularly busy."

In [52]:
# Create column “busiest_days" from "orders_day_of_week" to identify two busiest days (0, 1), slowest days (4, 3), or regularly busy (all other days).

result_2 = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_2.append("Busiest days")
  elif value == 4 or value == 3:
    result_2.append("Slowest days")
  else:
    result_2.append("Regularly busy")

In [53]:
result_2

['Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',

# Task Step 3 - Check values of the new "busiest days" column for accuracy with observations in markdown format.

In [54]:
# Create new 'busiest_days' column in ords_prods_merge to view results in context.

df_ords_prods_merge['busiest_days'] = result_2

In [55]:
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

In [57]:
# Check output of ords_prods_merge with new "busiest days" column.

df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  \
0   3139998      138    prior            28                   6   
1   1977647      138    prior            30                   6   
2    389851      709    prior             2                   0   
3    652770      764    prior             1                   3   
4   1813452      764    prior             3                   4   

   order_hour_of_day  days_since_prior_order  add_to_cart_order  reordered  \
0                 11                     3.0                  5          0   
1                 17                    20.0                  1          1   
2                 21                     6.0                 20          0   
3                 13                     NaN                 10          0   
4                 17                     9.0                 11          1   

  _merge    price_range_loc     busiest_day    busiest_days  
0   both  Mid-range product  Regularly busy  Regularly busy  
1   both  Mid-range product  Regularly busy  Regularly busy  
2   both  Mid-range product     Busiest day    Busiest days  
3   both  Mid-range product  Regularly busy    Slowest days  
4   both  Mid-range product      Least busy    Slowest days

#The total value counts for the ords_prods_merge dataframe is equal to the total sum of all groupings (i.e. "Regularly busy" + "Busiest days" + "Slowest days"). In addition, the listed sums for each label match the sum totals for the qualifying days of the week (e.g. Total "Busiest days" = 11864412, which matches orders_day_of_week 0 + 1, or 6204182 + 5660230).

# Task Step 4-Create a new column "busiest_period_of_day" to identify time periods “Most orders,” “Average orders,” and “Fewest orders.”

In [60]:
# Check value counts in "order_hour_of_day" column.

df_ords_prods_merge['order_hour_of_day'].value_counts()

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

The value counts listed above will be split into equal thirds for the following labels: "Most orders" = 10, 11, 14, 15, 13, 12, 16,9; "Fewest orders" = 23, 6, 0, 1, 5, 2, 4,3; and "Average orders" = all remaining values in order_hour_of_day.

In [64]:
# Create for-loop if statement labeling periods of time as “Most orders,” “Average orders,” and “Fewest orders.”

result_3 = []

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result_3.append("Most orders")
  elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
    result_3.append("Fewest orders")
  else:
    result_3.append("Average orders")

In [65]:
result_3

['Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Average orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most ord

In [66]:
# Create new column "busiest_period_of_day" in ords_prods_merge.

df_ords_prods_merge['busiest_period_of_day'] = result_3

# Task Step 5- Print the frequency for this new column.

In [67]:
# Print value counts in the "busiest_period_of_day" column.

df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

#  Step 7. Export dataframe as a pickle file to “Prepared Data” folder.

In [73]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))